### PR TITLE
[ir] Record the return types to a StructType

### PR DESCRIPTION
**Pull request by lin-hitonami**
Issue: #6983

### CODE REVIEW

File: taichi/program/callable.cpp [hunk 1]
```diff
@@ -31,6 +31,15 @@ int Callable::insert_texture_arg(const DataType &dt) {
   return (int)args.size() - 1;
 }
 
+void Callable::finalize_rets() {
+  std::vector<const Type *> types;
```
Comment (bobcao3) [inline taichi/program/callable.cpp:35]: nit:
```suggestion
types.reserve(rets.size());
```

File: taichi/program/callable.cpp [hunk 2]
```diff
@@ -31,6 +31,15 @@ int Callable::insert_texture_arg(const DataType &dt) {
...
+  for (auto &ret : rets) {
```
Comment (bobcao3) [inline taichi/program/callable.cpp:36]: tiny nit: `auto` does not deduce const (use const auto)

**Review by bobcao3 — APPROVED** [review]
LGTM

**Review by ailzhang — APPROVED** [review]
LGTM!

**Pull request merged**